tools/SampleAnalyzer/Process/Counter/CounterManager.h

```diff
@@ -35,80 +33,75 @@
-  // -------------------------------------------------------------
- public :
-
-  /// Constructor without argument 
-  CounterManager()
-  { }
-
-  /// Destructor
-  ~CounterManager()
-  { }
-
-  /// Initialize
-  void Initialize(const MAuint32& n)
-  { counters_.resize(n); }
-
-
-  // Specifying a cut name
-  void InitCut(const std::string myname)
-  {
-    Counter tmpcnt(myname);
-    counters_.push_back(tmpcnt);
-  }
-
-  /// Reset
-  void Reset()
-  { counters_.clear(); }
-
-  /// Overloading operator []
-  const Counter& operator[] (const MAuint32& index) const
-  { return counters_[index];}
-  Counter& operator[] (const MAuint32& index)
-  { return counters_[index];}
-
-  /// Incrementing the initial number of events
-  void IncrementNInitial(MAfloat32 weight=1.0)
-  { initial_.Increment(weight); }
-
-  /// Incrementing the initial number of events
-  Counter& GetInitial()
-  { return initial_; }
-  const Counter& GetInitial() const
-  { return initial_; }
-
-  /// Write the counters in a Text file
-  void Write_TextFormat(SAFWriter& output) const;
-
-  /// Write the counters in a ROOT file
-  //  void Write_RootFormat(TFile* output) const;
-
-  /// Finalizing
-  void Finalize()
-  { Reset(); }
-
-};
+    class CounterManager
+    {
+
+        // -------------------------------------------------------------
+        //                        data members
+        // -------------------------------------------------------------
+    private:
+        /// @brief intialisation indicator
+        MAbool initialised_;
+
+        // Collection of counters
+        std::vector<Counter> counters_;
+
+        // Initial number of events
+        Counter initial_;
+
+        // -------------------------------------------------------------
+        //                       method members
+        // -------------------------------------------------------------
+    public:
+        /// Constructor without argument
+        CounterManager() { initialised_ = false; }
+
+        /// Destructor
+        ~CounterManager() {}
+
+        /// Initialize
+        void Initialize(const MAuint32 &n) { counters_.resize(n); }
+
+        // Specifying a cut name
+        void InitCut(const std::string myname)
+        {
+            Counter tmpcnt(myname);
+            counters_.push_back(tmpcnt);
+        }
+
+        /// Reset
+        void Reset() { counters_.clear(); }
+
+        /// Overloading operator []
+        const Counter &operator[](const MAuint32 &index) const { return counters_[index]; }
+        Counter &operator[](const MAuint32 &index) { return counters_[index]; }
+
+        /// Incrementing the initial number of events
+        void IncrementNInitial(const WeightCollection &weight)
+        {
+            initial_.Increment(weight);
+            if (!initialised_)
+            {
+                for (auto &counter : counters_)
+                    counter.Initialise(weight);
+                initialised_ = true;
+            }
+        }
+
```

Comment:
I think there is a mistake on l85-91. The call to`Increment` should come **after** the initialisation test, and not before it (otherwise the code will try to fill weights in an unintialised counter).

Reply:
agreed!